ethcore/src/engines/clique/tests.rs

```diff
@@ -0,0 +1,773 @@
+			None
+		}).collect();
+		tags.sort();
+		tags
+	}
+
+	fn new_block_and_import(
+		&self,
+		block_type: CliqueBlockType,
+		last_header: &Header,
+		beneficary: Option<Address>,
+		signer: char,
+		) -> Result<OpenBlock, Error> {
+
+		let mut extra_data = vec![0; VANITY_LENGTH];
+		let mut seal = util::null_seal();
+		let last_hash = last_header.hash();
+
+		match block_type {
+			CliqueBlockType::Checkpoint => {
+				let signers = self.clique.state(&last_header).unwrap().signers().clone();
+				for signer in signers {
+					extra_data.extend(&*signer);
+				}
+			}
+			CliqueBlockType::Vote(v) => seal = v.as_rlp(),
+			CliqueBlockType::Empty => (),
+		};
+
+		let db = self.db.boxed_clone();
+
+		let mut block = OpenBlock::new(
+			&self.clique,
+			Default::default(),
+			false,
+			db,
+			&last_header.clone(),
+			Arc::new(vec![last_hash]),
+			beneficary.unwrap_or_default(),
+			(3141562.into(), 31415620.into()),
+			extra_data,
+			false,
+			&mut Vec::new().into_iter(),
+		).unwrap();
+
+		{
+			let difficulty = self.get_difficulty(block.header().number(), last_header, &self.signers[&signer].address());
+			let b = block.block_mut();
+			b.header.set_timestamp(last_header.timestamp() + self.clique.period);
+			b.header.set_difficulty(difficulty);
+			b.header.set_seal(seal);
+
+			let sign = ethkey::sign(self.signers[&signer].secret(), &b.header.hash()).unwrap();
+			let mut extra_data = b.header.extra_data().clone();
+			extra_data.extend_from_slice(&*sign);
+			b.header.set_extra_data(extra_data);
+		}
+
+		self.clique.verify_block_basic(block.header())?;
+		self.clique.verify_block_family(block.header(), &last_header)?;
+
+		Ok(block)
+	}
+}
+
+#[test]
+fn one_signer_with_no_votes() {
+	let tester = Tester::with(10, 1, vec![('A', true)]);
+
+	let empty_block = tester.new_block_and_import(CliqueBlockType::Empty, &tester.genesis, None, 'A').unwrap();
+
+	let tags = tester.tags_from_vec(&tester.clique_signers(&empty_block.header().hash()));
+	assert_eq!(&tags, &['A']);
+}
+
+#[test]
+fn one_signer_two_votes() {
+	let tester = Tester::with(10, 1, vec![('A', true), ('B', false), ('C', false)]);
+
+	// Add a vote for `B` signed by A
+	let vote = tester.new_block_and_import(CliqueBlockType::Vote(VoteType::Add), &tester.genesis,
+										   Some(tester.signers[&'B'].address()), 'A').unwrap();
+	let tags = tester.tags_from_vec(&tester.clique_signers(&vote.header().hash()));
+	assert_eq!(&tags, &['A', 'B']);
+
+	// Add a empty block signed by `B`
+	let empty = tester.new_block_and_import(CliqueBlockType::Empty, &vote.header(), None, 'B').unwrap();
+
+	// Add vote for `C` signed by A but should not be accepted
+	let vote = tester.new_block_and_import(CliqueBlockType::Vote(VoteType::Add), &empty.header(),
+										   Some(tester.signers[&'C'].address()), 'A').unwrap();
+
+	let tags = tester.tags_from_vec(&tester.clique_signers(&vote.header().hash()));
+	assert_eq!(&tags, &['A', 'B']);
+}
+
+#[test]
+fn two_signers_six_votes_deny_last() {
```

Comment:
These tests are quite horrible to read I'm sorry about that could probably make sense to implement json de-serialization instead....